### ISSUE TITLE
ConcurrentModificationException (GainAbilityControlledEffect and other)

### ISSUE THREAD

**Issue by catman123456** [magefree/mage#7170]
Playing with Korvold and this happen

Game exception occurred: java.util.ConcurrentModificationException
Server version: 1.4.45V1 (build: 2020-08-29 20:13)
java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1361)
java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:449)
mage.abilities.AbilitiesImpl.containsKey(AbilitiesImpl.java:275)
mage.game.permanent.PermanentImpl.addAbility(PermanentImpl.java:372)
mage.abilities.effects.common.continuous.GainAbilityControlledEffect.apply(GainAbilityControlledEffect.java:115)
mage.abilities.effects.ContinuousEffectImpl.apply(ContinuousEffectImpl.java:102)
mage.abilities.effects.ContinuousEffects.apply(ContinuousEffects.java:1044)
mage.game.GameState.applyEffects(GameState.java:595)
mage.game.GameImpl.applyEffects(GameImpl.java:1539)
mage.game.GameImpl.playPriority(GameImpl.java:1344)
mage.game.turn.Step.priority(Step.java:61)
mage.game.turn.Phase.playStep(Phase.java:184)
mage.game.turn.Phase.play(Phase.java:89)
mage.game.turn.Turn.play(Turn.java:126)
mage.game.GameImpl.playTurn(GameImpl.java:879)
mage.game.GameImpl.play(GameImpl.java:791)
mage.game.GameImpl.start(GameImpl.java:776)
mage.server.game.GameWorker.call(GameWorker.java:34)
java.util.concurrent.FutureTask.run(FutureTask.java:266)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)

**Comment by LevelX2** [comment]
Probably no direct relation to Korvold, Fae-Cursed King.

It would be helpful to describe the exact situation on battlefield/stack while this happened.

**Comment by JayDi85** [comment]
I can't see that type of errors in beta logs so it must be fixed with new version.

**Comment by JayDi85** [comment]
Origin of the bug: abilities list somehow changed while cycling in the same list. Two possible source of the problem:
1. Another thread changed the source list (need research, maybe disconnects/timeouts/rollback from another player can cause it);
2. Added new ability to the same list (see below).

Example for second use case

`GainAbilityControlledEffect` -- permanent target itself and try to add ability to itself. So it can raise the error in theory. Need real life example. Another possible reason: `LKI` somehow saved a real object instead object.copy() (or that object miss abilities list copy).
![shot_210213_155424](https://user-images.githubusercontent.com/8344157/107849416-7833e400-6e14-11eb-924d-5126abe65f39.png)

**Comment by JayDi85** [comment]
Server logs from human games -- CME with ContinuousEffectsList:
```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1445)
	at java.util.HashMap$EntryIterator.next(HashMap.java:1479)
	at java.util.HashMap$EntryIterator.next(HashMap.java:1477)
	at mage.abilities.effects.ContinuousEffectsList.<init>(ContinuousEffectsList.java:33)
	at mage.abilities.effects.ContinuousEffectsList.copy(ContinuousEffectsList.java:43)
	at mage.abilities.effects.ContinuousEffects.<init>(ContinuousEffects.java:80)
	at mage.abilities.effects.ContinuousEffects.copy(ContinuousEffects.java:108)
	at mage.game.GameState.<init>(GameState.java:158)
	at mage.game.GameState.copy(GameState.java:246)
	at mage.game.GameImpl.rollbackTurnsExecution(GameImpl.java:3304)
	at mage.game.GameImpl.playTurn(GameImpl.java:891)
	at mage.game.GameImpl.play(GameImpl.java:812)
	at mage.game.GameImpl.start(GameImpl.java:797)
	at mage.server.game.GameWorker.call(GameWorker.java:34)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by JayDi85** [comment]
Fixed in 1664ee01cf30af2e85d725d63f22808ffdf2d61f
